pkg/runtime/sequencer/cache.go

```diff
@@ -2,15 +2,151 @@ package sequencer
+	"sync"
 )
 
+const defaultSize = 1000
+const defaultLimit = 700
+
+// DoubleBuffer is double segment id buffer.
+// There are two buffers in DoubleBuffer: InUseBuffer is in use, BackUpBuffer is a backup buffer.
+// Their default capacity is 1000. When the InUseBuffer usage exceeds 30%, the BackUpBuffer will be initialized.
+// When InUseBuffer is used up, swap them.
+type DoubleBuffer struct {
+	Key          string
+	Size         int
+	InUseBuffer  *Buffer
+	BackUpBuffer *Buffer
+	lock         sync.Mutex
+	Store        sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:   key,
+		Size:  defaultSize,
+		Store: store,
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.InUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	if d.InUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: InUseBuffer nil ")
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	next := d.InUseBuffer.from
+	d.InUseBuffer.from++
+
+	//check swap
+	if d.InUseBuffer.from > d.InUseBuffer.to {
+		d.swap()
+	}
+
+	//when InUseBuffer id more than half used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.InUseBuffer.to-d.InUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+
+			buffer, err := d.getNewBuffer()
+			if err != nil {
+				log.DefaultLogger.Errorf("[DoubleBuffer] [getNewBuffer] error: %v", err)
+			}
+			d.BackUpBuffer = buffer
+
+		}, nil)
+	}
+
+	return next, nil
+}
+
+//swap InUseBuffer and BackUpBuffer, must be locked
+func (d *DoubleBuffer) swap() {
+	d.InUseBuffer = d.BackUpBuffer
+	d.BackUpBuffer = nil
+}
+
+//getNewBuffer return a new segment
+func (d *DoubleBuffer) getNewBuffer() (*Buffer, error) {
+	support, result, err := d.Store.GetSegment(&sequencer.GetSegmentRequest{
+		Key:  d.Key,
+		Size: d.Size,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if !support {
+		return nil, errors.New("[DoubleBuffer] unSupport Segment id")
+	}
+	return &Buffer{
+		from: result.From,
+		to:   result.To,
+	}, nil
+}
+
+var BufferCatch sync.Map
+
 func GetNextIdFromCache(ctx context.Context, store sequencer.Store, req *sequencer.GetNextIdRequest) (bool, int64, error) {
-	// TODO
-	//	1. check cache
-	//	2. load cache with lock
-	//	2.1. return if not support
-	//	2.2. add segment into cache
-	//	2.3. return result
-	return false, 0, nil
+
+	//check support
+	support, _, _ := store.GetSegment(&sequencer.GetSegmentRequest{
+		Key:  req.Key,
+		Size: 0,
+	})
+
+	if !support {
+		return false, 0, nil
+	}
+
+	actual, ok := BufferCatch.LoadOrStore(req.Key, NewDoubleBuffer(req.Key, store))
+
+	d := actual.(*DoubleBuffer)
+	//load fail, should init buffer
+	if !ok {
+		err := d.init()
```

Comment:
concurrent GetNextIdFromCache may got not inited DoubleBuffer?

Reply:
@ZLBer  yes, only return error now

Reply:
then getId will fail, which should be prevented

Reply:
@MoonShining  I'm not very familiar with the features of golang. maybe we can  block other goroutine with  `sync.WaitGroup` ?

Reply:
use rwmutex and a map is simpler